examples/torch/introduction_to_torch_geometric_for_atomistic_machine_learning.py

```diff
@@ -0,0 +1,230 @@
+# 
+# Source: `When Work Matters: Transforming Classical Network Structures to
+# Graph CNN <https://arxiv.org/abs/1807.02653>`__
+# 
+
+from torch_geometric.datasets import TUDataset
+
+dataset = TUDataset(root='/tmp/ENZYMES', name='ENZYMES', use_node_attr=True)
+
+dataset # Contains 600 graphs
+
+
+######################################################################
+# Elements of the dataset are Data objects. The first data object in the
+# dataset has 37 nodes, 168 edges, and 21 features for each node. The
+# graph is also labeled with the class 5, that is stored in the ``y``
+# attribute of the data object.
+# 
+
+
+######################################################################
+# The dataset also has metadata stored in form of attributes of the
+# dataset object. The ``num_classes`` attribute contains the number of
+# classes in the dataset, and the ``num_node_features`` attribute contains
+# the number of node features in the dataset, and the
+# ``num_edge_features`` attribute contains the number of edge features in
+# the dataset.
+# 
+# These options will not necessarily be presented in all datasets. For
+# example, the ``Cora`` dataset does not have edge features, and the
+# ``QM9`` dataset does not have node labels.
+# 
+
+print(f"Total number of entries: {len(dataset)}")
+print(f'Number of node features: {dataset.num_node_features}')
+print(f'Number of edge features: {dataset.num_edge_features}')
+print(f'Number of classes: {dataset.num_classes}')
+
+
+######################################################################
+# For conveniece, the dataset may be shuffled by calling the ``shuffle()``
+# method of the dataset object. Please note, that it works out-of-place.
+# 
+
+dataset = dataset.shuffle()
+
+
+######################################################################
+# 3. Mini-batches
+# ---------------
+# 
+
+
+######################################################################
+# One of the main issued with the graph data is that the graphs may have
+# different number of nodes and edges. This feature requires to use a
+# specific way of joining the graphs into mini-batches. The PyTorch
+# Geometric library provides a special class for this purpose -
+# ``torch_geometric.loader.DataLoader``. It works in the same way as the
+# original PyTorch ``DataLoader`` class, but it joins the graphs in a way
+# that allows to use them in the graph neural networks.
+# 
+
+
+######################################################################
+# The way it works is that it joins all graphs in the mini-batch into a
+# single large graph. The connectivity information for this new graph is
+# stored in a new block diagonal adjacency matrix. The node and edge
+# features are stored in a single large node feature tensor and edge
+# feature tensor, respectively.
+# 
+# I.e., if we had a graph :math:`G_1 = (X_1, A_1)` and graph
+# :math:`G_2 = (X_2, A_2)`, then uniting them into a Batch will result in
```

Comment:
I feel like this part is unnecessary abstract explanation for "We stack graph data together and store an index for each graph/structure".